gpcontrib/arenadata_toolkit/src/bloom_set.c

```diff
@@ -0,0 +1,280 @@
+
+/* simple linear search, probably should be reworked (depends on target dbs count) */
+static bloom_entry_t * find_bloom_entry(bloom_set_t * bloom_set, Oid dbid, bool *found)
+{
+	bloom_entry_t *bloom_entry;
+	int			i;
+
+	*found = false;
+
+	for (i = 0; i < bloom_set->bloom_count; i++)
+	{
+		bloom_entry = BLOOM_ENTRY_GET(bloom_set, i);
+		if (bloom_entry->dbid == dbid || bloom_entry->dbid == InvalidOid)
+			break;
+	}
+
+	if (i == bloom_set->bloom_count)
+		return NULL;
+
+	if (bloom_entry->dbid != InvalidOid)
+		*found = true;
+
+	return bloom_entry;
+}
+
+/* bind not used filter to given dbid */
+bool
+bloom_set_bind(bloom_set_t * bloom_set, Oid dbid)
+{
+	bloom_entry_t *bloom_entry;
+	bool		found;
+
+	LWLockAcquire(bloom_set->lock, LW_SHARED);
+	bloom_entry = find_bloom_entry(bloom_set, dbid, &found);
+	LWLockRelease(bloom_set->lock);
+
+	if (found)
+		return true;
+	else if (!bloom_entry)
+	{
+		return false;
+	}
+
+	LWLockAcquire(bloom_set->lock, LW_EXCLUSIVE);
+	bloom_entry = find_bloom_entry(bloom_set, dbid, &found);
+	if (bloom_entry && !found)
+		LWLockAcquire(bloom_entry->lock, LW_EXCLUSIVE);
+	LWLockRelease(bloom_set->lock);
+
+	if (!bloom_entry)
+	{
+		elog(WARNING, "Our bloom filter was stolen :(");
+		return false;
+	}
+
+	if (!found)
+	{
+		bloom_entry->dbid = dbid;
+		LWLockRelease(bloom_entry->lock);
+		elog(DEBUG1, "Bloom binded %d", dbid);
+	}
+
+
+	return true;
+}
+
+bool
+bloom_set_trigger_bits(bloom_set_t * bloom_set, Oid dbid, bool on)
+{
+	bloom_entry_t *bloom_entry;
+	bool		found;
+
+	LWLockAcquire(bloom_set->lock, LW_SHARED);
+	{
+		bloom_entry = find_bloom_entry(bloom_set, dbid, &found);
+	}
+	LWLockRelease(bloom_set->lock);
+
+	if (!found)
+	{
+		elog(LOG, "[arenadata toolkit] tracking_initial_snapshot Bloom filter not found");
+		return false;
+	}
+
+	LWLockAcquire(bloom_entry->lock, LW_EXCLUSIVE);
+	{
+		if (on)
+			bloom_set_all(&bloom_entry->bloom);
+		else
+			bloom_clear(&bloom_entry->bloom);
+	}
+	LWLockRelease(bloom_entry->lock);
+	return true;
+}
+
+/* unbind used filter by given dbid */
+void
+bloom_set_unbind(bloom_set_t * bloom_set, Oid dbid)
+{
+	bloom_entry_t *bloom_entry;
+	bool		found;
+
+	LWLockAcquire(bloom_set->lock, LW_SHARED);
+	bloom_entry = find_bloom_entry(bloom_set, dbid, &found);
+	LWLockRelease(bloom_set->lock);
+
+	if (!found)
+		return;
+
+	LWLockAcquire(bloom_set->lock, LW_EXCLUSIVE);
+	bloom_entry = find_bloom_entry(bloom_set, dbid, &found);
+	if (bloom_entry && found)
+		LWLockAcquire(bloom_entry->lock, LW_EXCLUSIVE);
+	LWLockRelease(bloom_set->lock);
+
+	if (found)
+	{
+		bloom_entry->dbid = InvalidOid;
+		bloom_clear(&bloom_entry->bloom);
+		LWLockRelease(bloom_entry->lock);
+		elog(DEBUG1, "Bloom unbinded %d", dbid);
+	}
+}
+
+uint64_t
+bloom_set_calc_hash(const void *buf, size_t len)
+{
+	struct wide_hash
+	{
+		uint64_t	i1;
+		uint64_t	i2;
+	};
+	struct wide_hash w_hash;
+	bool		hash_res = pg_md5_binary(buf, len, &w_hash);
+
+	Assert(hash_res);
+
+	return w_hash.i1 ^ w_hash.i2;
```

Comment:
Can this be simplified?
```suggestion
	uint64_t	md5[2];
	bool		hash_res = pg_md5_binary(buf, len, md5);
	Assert(hash_res);
	return md5[0] ^ md5[1];
```

Reply:
Changed the approach